File: sach/de3306.cpp
//DE3306.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
int F[100001];
const int MOD = 1e9 + 7;
int main(){
    int n;
    cin >> n;
    set<string> s;
    for (int i = 0; i < n; i++) {
        string p;
        cin >> p;
        s.insert(p);
    }
    string x;
    cin >> x;
    int m = x.length();
    F[0] = 1;
    for (int i = 1; i <= m; i++) {
        F[i] = 0;
        for (int l = 1; l <= 100 && i - l >= 0; l++){
            int j = i - l;
            if (s.find(x.substr(j, l)) != s.end())
                F[i] = (F[i] + F[j]) % MOD;
        }
    }
    cout << F[m];
    return 0;
}

File: sach/de3306.py
//DE3306.PY - Tran Huu Nam - hsgtin.vn - 2024
MOD = int(1e9 + 7)
def main():
    n = int(input())
    s = set()
    for _ in range(n):
        p = input().strip()
        s.add(p)

    x = input().strip()
    m = len(x)
    F = [0] * (m + 1)
    F[0] = 1

    for i in range(1, m + 1):
        for l in range(1, 101):
            j = i - l
            if j >= 0 and x[j:i] in s:
                F[i] = (F[i] + F[j]) % MOD

    print(F[m])
if __name__ == "__main__":
    main()

